auth/service/api/v1/devicetokens.go

```diff
@@ -0,0 +1,96 @@
+package v1
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/ant0ine/go-json-rest/rest"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/request"
+	"github.com/tidepool-org/platform/service/api"
+)
+
+func (r *Router) DeviceTokensRoutes() []*rest.Route {
+	return []*rest.Route{
+		rest.Post("/v1/users/:userId/device_tokens", api.RequireAuth(r.UpsertDeviceToken)),
+	}
+}
+
+func (r *Router) UpsertDeviceToken(res rest.ResponseWriter, req *rest.Request) {
+	responder := request.MustNewResponder(res, req)
+	ctx := req.Request.Context()
+	authDetails := request.GetAuthDetails(ctx)
+	repo := r.AuthStore().NewDeviceTokenRepository()
+
+	if err := checkAuthentication(authDetails); err != nil {
+		log.Printf("checkAuth failed: %+v", authDetails)
+		responder.Error(http.StatusUnauthorized, err)
+		return
+	}
+
+	if err := checkUserIDConsistency(authDetails, req.PathParam("userId")); err != nil {
+		log.Printf("checkUserIDConsistency failed: %+v %q", authDetails, req.PathParam("userID"))
+		responder.Error(http.StatusForbidden, err)
+		return
+	}
+
+	deviceToken := devicetokens.DeviceToken{}
+	if err := request.DecodeRequestBody(req.Request, &deviceToken); err != nil {
+		responder.Error(http.StatusBadRequest, err)
+		return
+	}
+
+	userID := userIDWithServiceFallback(authDetails, req.PathParam("userId"))
+	doc := devicetokens.NewDocument(userID, deviceToken)
+	if err := repo.Upsert(ctx, doc); err != nil {
+		responder.Error(http.StatusInternalServerError, err)
+		return
+	}
+}
+
+var ErrUnauthorized = fmt.Errorf("unauthorized")
+
+// checkAuthentication ensures that the request has an authentication token.
+func checkAuthentication(details request.AuthDetails) error {
+	if details.Token() == "" {
+		return ErrUnauthorized
+	}
+	if details.IsUser() {
+		return nil
+	}
+	if details.IsService() {
+		return nil
+	}
+	return ErrUnauthorized
```

Comment:
> Hmm, looking at this code, I'm pretty sure this isn't necessary (though my memory is a bit fuzzy here). The `api.RequireAuth` wrapper on the handler above should require and enforce proper authorization, so you shouldn't need this function at all.

You're not wrong, but there's no way for this controller to know that api.RequireAuth has wrapped it though, is there? I'd rather be safe than sorry. I'm a fan of controllers knowing and being explicit about these things, preferably in their definitions where you can't miss it. That way, if it gets moved or copied to some other service, or some other path, there aren't any worries about api.RequireAuth being accidentally left off.

> Also, FYI, the `details.Token()` will incorrectly fail if the API call is a service-to-service all using a service secret.

It's correct for this to fail if called by a service. They're not allowed to call this. It has to come from a Tidepool app on a client device. If this comes from one of our services, something is very wrong.

Reply:
Well, the `api.RequireAuth` is in the same file (at the top where the routes are defined) so I'm not sure how more explicit we need to be.

Also, I'm confused. Lines 63-64:
```
	if details.IsService() {
		return nil
	}
```
says that if its a service then it is okay.

And, if a service isn't okay, then the `api.RequireAuth` should be changed to `api.RequireUser`, which will do what you want.

Reply:
Thanks for that, updated.

> Well, the api.RequireAuth is in the same file (at the top where the routes are defined) so I'm not sure how more explicit we need to be.

It's probably scar tissue on my part from previous projects. 🤷🏻 Relying on middleware for sensitive authentication/authorization has bitten me in the past when people moved or copied code. So I've often found it better to have those checks directly inside each handler, so that they go with it, no matter where it ends up. Having now typed that out and re-read it a few times, it sounds a little overboard, even to me. :)

For another example, I feel very nervous that we allow a userId to be provided in our paths, when there's a userId in the authentication token. I'm not aware of any other code that even checks if they differ..? I'd prefer to have one and only one source for important things like a userId.

I can understand that putting the userId in the path could be better (or even required?) for service tokens that aren't tied to a userId. But it seems to me that the right way to deal with this is to have two separate endpoints, one usable only with a user token and that doesn't include a userid in the path, and a second endpoint that's usable only with a service token that requires a userid in the path. That way we completely eliminate the possibility of a userId from a path conflicting with a userId from a token. It would also eliminate the possibility of a user posting with their own token, but with a path that includes another user's userId, in order to modify another user's data. I feel like this is a hidden time-bomb that will bite us eventually.

> Also, I'm confused. Lines 63-64:

This function's goal was to double check that authentication had been done. Not authorization. ;) But as you pointed out, RequireUser should do both of these things for me, and that's the convention in platform, so I'll moving to that.

Reply:
I agree with your statement about trying to eliminate the path vs. auth token user id discrepancy risk, but there are use cases where they will be different by design. 

For example, one user can upload data on behalf of another user, such as a clinician uploading data for a PwD. The path user id is for the PwD and the auth token user id is for the clinician. The code will ensure that the auth token user has permissions to operate on the path user's data. Here's an example when uploading data - https://github.com/tidepool-org/platform/blob/9f67aa8e4c2ed24d4a584c3f6b34a17906c659a5/data/service/api/v1/datasets_data_create.go/#L43-L59